### ISSUE TITLE
libgccjit DLL and the import library shouldn't export strsignal

### ISSUE THREAD

**Issue by pcrama** [msys2/MINGW-packages#7484]
## Describe the issue

I'm trying to build the emacs feature/native-comp branch and while trying to sort out linker issues, I was told by Eli Zaretskii to file this issue (see https://debbugs.gnu.org/cgi/bugreport.cgi?bug=45303#104)

[I wrote]
> strsignal is in libgccjit.dll.a but not in libgccjit-0.dll:
> $ for f in /mingw64/bin/libgccjit-0.dll /mingw64/lib/libgccjit.dll.a ; do (nm -A "$f" | grep strsignal) || echo "Not
> found in $f" ; done
> Not found in /mingw64/bin/libgccjit-0.dll
> C:/msys64/mingw64/lib/libgccjit.dll.a:d025724.o:0000000000000000 I __imp_strsignal
> C:/msys64/mingw64/lib/libgccjit.dll.a:d025724.o:0000000000000000 T strsignal

[He wrote]
> This probably means the configure-time test for strsignal should be moved to before the test for libgccjit.
>
> But in general, I'd suggest to file a bug report with MSYS2 folks: the DLL and the import library shouldn't export strsignal.

## Steps to Reproduce the Problem

1. Run `pacman -S --needed base-devel mingw-w64-x86_64-toolchain mingw-w64-x86_64-libgccjit`
2. Run `(nm -A /mingw64/lib/libgccjit.dll.a | grep strsignal) || echo "Everything OK, strsignal not exported"`
3. Observe this output:
```
C:/msys64/mingw64/lib/libgccjit.dll.a:d025724.o:0000000000000000 I __imp_strsignal
C:/msys64/mingw64/lib/libgccjit.dll.a:d025724.o:0000000000000000 T strsignal
```
4. I expected just `Everything OK, strsignal not exported` to be printed.

## Context

I'm trying to build Emacs (the feature/native-comp branch) and had many problems (see https://debbugs.gnu.org/cgi/bugreport.cgi?bug=45303 for full details and discussion).  One of the issues that was raised there is that the configure script was fooled into believing that `strsignal` was available, later in the build process linking fails partly because `strsignal` is not available (due to another issue I do not understand, the final linking does not include libgccjit).  But my understanding is that while indeed the emacs build process should be improved to include the `-lgccjit` flag at link time, it is still not libgccjit's job to export that identifier.

This is the full build process I am using:
```
  pacman -S --needed base-devel \
    mingw-w64-x86_64-toolchain \
    mingw-w64-x86_64-xpm-nox \
    mingw-w64-x86_64-libtiff \
    mingw-w64-x86_64-giflib \
    mingw-w64-x86_64-libpng \
    mingw-w64-x86_64-libjpeg-turbo \
    mingw-w64-x86_64-librsvg \
    mingw-w64-x86_64-lcms2 \
    mingw-w64-x86_64-jansson \
    mingw-w64-x86_64-libxml2 \
    mingw-w64-x86_64-gnutls \
    mingw-w64-x86_64-zlib \
    mingw-w64-x86_64-harfbuzz \
    mingw-w64-x86_64-libgccjit
  PROCESSORS_TO_USE="3"
  EMACS_VERSION=emacs-native-comp
  mkdir -p /tmp/$EMACS_VERSION
  cd /tmp/$EMACS_VERSION
  git clone https://git.savannah.gnu.org/git/emacs.git
  git checkout 682bd303470d4a0fcd2690aff6aa58fb720a8d41
  cd emacs
  ./autogen.sh
  # edit nt/epaths.nt to add PATH_REL_LOADSEARCH:
  grep -q PATH_REL_LOADSEARCH nt/epaths.nt || echo '#define PATH_REL_LOADSEARCH ""' >> nt/epaths.nt
  # patch to look for libgccjit-0.dll instead of libgcc.dll?  lisp/term/w32-win.el & src/emacs.c
  sed -i -e 's/libgccjit.dll/libgccjit-0.dll/' lisp/term/w32-win.el
  sed -i -e 's/libgccjit.dll/libgccjit-0.dll/' src/emacs.c
  # patch to avoid gcc_jit_global_set_initializer (crashes on my machine...; it
  # seems there is an interaction with the #pragma and the rest of the parsing
  # because the statement is incomplete?) and to adapt to (new) 5th parameter to
  # directory-files
  sed -i -e '/if (gcc_jit_global_set_initializer)/,/{/ {
               /#pragma GCC diagnostic pop/d
               /{/a #pragma GCC diagnostic pop
             }' \
         -e '/internal_condition_case_4/,/FOR_EACH/ {
               s/internal_condition_case_4/internal_condition_case_5/
               s/Qt, return_nil);/Qnil, Qt, return_nil);/
             }' \
         src/comp.c
  sed -i -e '/extern Lisp_Object internal_condition_case_4/a extern Lisp_Object internal_condition_case_5 (Lisp_Object (*) (Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object), Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object, Lisp_Object (*) (Lisp_Object));' src/lisp.h
  sed -i -e '/Like internal_condition_case_1 but call BFUN with ARG1, ARG2, ARG3, ARG4 as/ {
               s/ARG4 as/ARG4, ARG5 as/
               a    its arguments.  */
               a Lisp_Object
               a internal_condition_case_5 (Lisp_Object (*bfun) (Lisp_Object, Lisp_Object,
               a                                                 Lisp_Object, Lisp_Object,
               a                                                 Lisp_Object),
               a                            Lisp_Object arg1, Lisp_Object arg2,
               a                            Lisp_Object arg3, Lisp_Object arg4,
               a                            Lisp_Object arg5,
               a                            Lisp_Object handlers,
               a                            Lisp_Object (*hfun) (Lisp_Object))
               a {
               a   struct handler *c = push_handler (handlers, CONDITION_CASE);
               a   if (sys_setjmp (c->jmp))
               a     {
               a       Lisp_Object val = handlerlist->val;
               a       clobbered_eassert (handlerlist == c);
               a       handlerlist = handlerlist->next;
               a       return hfun (val);
               a     }
               a   else
               a     {
               a       Lisp_Object val = bfun (arg1, arg2, arg3, arg4, arg5);
               a       eassert (handlerlist == c);
               a       handlerlist = c->next;
               a       return val;
               a     }
               a }
               a /* Like internal_condition_case_1 but call BFUN with ARG1, ARG2, ARG3, ARG4 as
             }' src/eval.c
  sed -i -e '/PATH_EXEC, 0);/ {
               s/.*/#ifdef WINDOWSNT/
               a /* On MS-Windows, PATH_EXEC normally starts with a literal
               a    "%emacs_dir%", so it will never work without some tweaking. */
               a w32_relocate (PATH_EXEC),
               a #else
               a PATH_EXEC,
               a #endif
               a 0);
             }' src/callproc.c
  mkdir -p ../build
  cd ../build
  ../emacs/configure \
        --with-xml2 \
        --without-pop \
        --prefix="/tmp/$EMACS_VERSION/dest" \
        --without-compress-install \
        --without-dbus \
        --with-nativecomp \
        --with-modules 'CFLAGS=-O2 -g3' 'LIBGCCJIT=-lz -lgccjit'
  sed -i -e 's/^LIBGCCJIT = *$/LIBGCCJIT = -lz -lgccjit/' src/Makefile
  make -j"$PROCESSORS_TO_USE"
  make install
```

**Comment by lazka** [comment]
I'm afraid no one here knows much about libgccjit, we just ship it as is (via a backport.. maybe there have been more upstream fixes?)

**Comment by clockoon** [comment]
For a temporary solution, you can re-generate `libgccjit.dll.a` file using `gendef` and `dlltool`.

**Comment by dr-scsi** [comment]
@lazka - native-comp has landed in Emacs and is included in 28.1 which builds on Windows with Msys2.  I think this bug can be closed.
